pkg/topo/topo.go

```diff
@@ -2,70 +2,174 @@
 	"google.golang.org/grpc/status"
-	"io"
+
+	"github.com/onosproject/onos-ric-sdk-go/pkg/utils/creds"
+	"google.golang.org/grpc/credentials"
+
+	topoapi "github.com/onosproject/onos-api/go/onos/topo"
+
+	"github.com/onosproject/onos-lib-go/pkg/logging"
+
+	"time"
+
+	"github.com/onosproject/onos-ric-sdk-go/pkg/topo/connection"
+
+	"github.com/onosproject/onos-lib-go/pkg/southbound"
+	"google.golang.org/grpc"
 )
 
-var log = logging.GetLogger("topo", "watch")
+var log = logging.GetLogger("topo")
 
-// Watch provides a simple facility for the application to watch for changes in the topology.
-func (c *topoClient) Watch(ctx context.Context, ch chan<- topoapi.Event, filter ...FilterOption) error {
-	// Issue a watch request with replay and passing through any service-side filter criteria, i.e.
-	// those supported by the topo.Watch method (TBD)
-	stream, err := c.client.Watch(ctx, &topoapi.WatchRequest{})
-	if err != nil {
-		log.Error("Unable to issue topology watch request", err)
-		stat, ok := status.FromError(err)
-		if ok {
-			return errors.FromStatus(stat)
+// Client is a topo SDK client
+type Client interface {
+
+	// Update updates a topo object
+	Update(ctx context.Context, object *topoapi.Object) error
+
+	// Get gets a topo object with a given ID
+	Get(ctx context.Context, id topoapi.ID) (*topoapi.Object, error)
+
+	// Watch provides a simple facility for the application to watch for changes in the topology
+	Watch(ctx context.Context, ch chan<- topoapi.Event, opts ...WatchOption) error
+
+	// List of topo objects
+	List(ctx context.Context, opts ...ListOption) ([]topoapi.Object, error)
+}
+
+// NewClient creates a new topo client
+func NewClient(opts ...Option) (Client, error) {
+	clientOptions := Options{}
+
+	for _, opt := range opts {
+		opt.apply(&clientOptions)
+	}
+
+	if clientOptions.Service.Host == "" || clientOptions.Service.Port == 0 {
+		clientOptions.Service.Host = DefaultServiceHost
+		clientOptions.Service.Port = DefaultServicePort
+	}
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithStreamInterceptor(southbound.RetryingStreamClientInterceptor(100 * time.Millisecond)),
+		grpc.WithUnaryInterceptor(southbound.RetryingUnaryClientInterceptor()),
+	}
+	if clientOptions.Service.Insecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	} else {
+		tlsConfig, err := creds.GetClientCredentials()
+		if err != nil {
+			log.Warn(err)
+			return nil, err
 		}
-		return err
+
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	}
+	conns := connection.NewManager()
+	conn, err := conns.Connect(clientOptions.Service.GetAddress(), dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	cl := topoapi.NewTopoClient(conn)
+
+	return &topo{
+		client: cl,
+	}, nil
+}
+
+// topo is the topo client
+type topo struct {
+	client topoapi.TopoClient
+}
+
+// Update updates a given topo object
+func (t *topo) Update(ctx context.Context, object *topoapi.Object) error {
```

Comment:
This topology API uses optimistic locking via the revision on the topo `Object`. When we update the object, the revision will change. The pointer passed in to this method should be updated with the updated object (returned in the update response).